pkg/controller/master/virtualmachine/vmi_network_controller.go

```diff
@@ -129,3 +202,65 @@ func patchVMRevision(vm *kubevirtv1.VirtualMachine) ([]byte, error) {
 	patch, err := json.Marshal(objCopy)
 	return patch, err
 }
+*/
+
+// clone of patchVMRevision but works upon VirtualMachineSpec
+func patchVMRevisionViaVMSpec(vmSpec *kubevirtv1.VirtualMachineSpec) ([]byte, error) {
+	vmBytes, err := json.Marshal(vmSpec)
+	if err != nil {
+		return nil, err
+	}
+	var raw map[string]interface{}
+	err = json.Unmarshal(vmBytes, &raw)
+	if err != nil {
+		return nil, err
+	}
+	objCopy := make(map[string]interface{})
+	objCopy["spec"] = raw
+	patch, err := json.Marshal(objCopy)
+	return patch, err
+}
+
+func (h *VMNetworkController) updateVM(vmi *kubevirtv1.VirtualMachineInstance) (*kubevirtv1.VirtualMachineInstance, error) {
+	vm, err := h.vmCache.Get(vmi.Namespace, vmi.Name)
+	if err != nil {
+		return vmi, err
+	}
+
+	if vm == nil || vm.DeletionTimestamp != nil {
+		return vmi, nil
+	}
+
+	// already synced
+	if vm.Generation == vm.Status.ObservedGeneration {
+		return vmi, nil
+	}
+
+	crObj, err := h.crCache.Get(vmi.Namespace, vmi.Status.VirtualMachineRevisionName)
+	if err != nil {
+		return vmi, fmt.Errorf("failed to fetch controllerRevision %s/%s: %w", vmi.Namespace, vmi.Status.VirtualMachineRevisionName, err)
+	}
+
+	revisionSpec := &vmwatch.VirtualMachineRevisionData{}
+	if err = json.Unmarshal(crObj.Data.Raw, revisionSpec); err != nil {
+		return vmi, fmt.Errorf("failed to unmarshal controllerRevision %s/%s: %w", crObj.Namespace, crObj.Name, err)
+	}
+
+	// if there are other changes than MAC, skip
+	if !equality.Semantic.DeepEqual(revisionSpec.Spec, vm.Spec) {
+		return vmi, nil
+	}
+
+	// forcely update vm, then upstream controller will sync the Generation and ObservedGeneration
+	vmCopy := vm.DeepCopy()
+	if vmCopy.Annotations == nil {
+		vmCopy.Annotations = make(map[string]string, 1)
+	}
+	vmCopy.Annotations[util.AnnotationTimestamp] = time.Now().Format(time.RFC3339)
```

Comment:
is the annotationTimestamp used by the UI? I see no actual logic implemented in the controllers to actually perform additional actions when this annotation is set.

Reply:
this `annotationTimestamp` has two usages:

(1) Let kubevirt's controller be triggered, in this context

(2) UI has a websocket watching VM object (not all realted vmi, vmim...), when VM object has change, it will be notified, and then render the menu, e.g., the vmim controller call below function when vmim object status changes, it helps UI to update `migrate/abortMigration` button.

https://github.com/harvester/harvester/blob/b4b8aeb7ba260e8d01d306cf316d6611b3d01d53/pkg/controller/master/migration/vmim_controller.go#L86

https://github.com/harvester/harvester/blob/b4b8aeb7ba260e8d01d306cf316d6611b3d01d53/pkg/controller/master/migration/vmim_controller.go#L237

Reply:
should have known we were updating the vm object and not the vmi object. if we could please address the feedback rest looks good.